File: Java/백준/Gold/9663. N－Queen/N－Queen.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.io.IOException;

public class Main {
    static int N;
    static int[] board;
    static int count = 0;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        N = Integer.parseInt(br.readLine());
        board = new int[N];
        dfs(0);
        System.out.println(count);
    }

    static void dfs(int depth) {
        if (depth == N) {
            count++;
            return;
        }
        for (int i = 0; i < N; i++) {
            board[depth] = i;
            if (isPossible(depth)) {
                dfs(depth + 1);
            }
        }
    }

    static boolean isPossible(int col) {
        for (int i = 0; i < col; i++) {
            if (board[col] == board[i]) return false;
            if (Math.abs(col - i) == Math.abs(board[col] - board[i])) return false;
        }
        return true;
    }
}

File: C11/백준/Gold/9663. N－Queen/N－Queen.c
#include <stdio.h>
#include <stdlib.h>

int N;
int count = 0;
int *board;

int isPossible(int col) {
    for (int i = 0; i < col; i++) {
        if (board[col] == board[i]) return 0;
        if (abs(col - i) == abs(board[col] - board[i])) return 0;
    }
    return 1;
}

void dfs(int depth) {
    if (depth == N) {
        count++;
        return;
    }
    for (int i = 0; i < N; i++) {
        board[depth] = i;
        if (isPossible(depth)) {
            dfs(depth + 1);
        }
    }
}

int main(void) {
    scanf("%d", &N);
    board = (int*)malloc(sizeof(int) * N);
    dfs(0);
    printf("%d\n", count);
    free(board);
    return 0;
}